pipelined/bevy_hdr/src/bloom.wgsl

```diff
@@ -0,0 +1,164 @@
+struct VertexOutput {
+	[[builtin(position)]] position: vec4<f32>;
+	[[location(0)]] uv: vec2<f32>;
+};
+
+var vertices: array<vec2<f32>, 3> = array<vec2<f32>, 3>(
+	vec2<f32>(-1.0, -1.0),
+	vec2<f32>(3.0, -1.0),
+	vec2<f32>(-1.0, 3.0),
+);
+
+[[stage(vertex)]]
+fn vertex([[builtin(vertex_index)]] idx: u32) -> VertexOutput {
+	var out: VertexOutput;
+
+	out.position = vec4<f32>(vertices[idx], 0.0, 1.0);
+	out.uv = vertices[idx] * vec2<f32>(0.5, -0.5);
+	out.uv = out.uv + 0.5;
+
+	return out;
+}
+
+[[block]]
+struct Uniforms {
+	threshold: f32;
+	knee: f32;
+	scale: f32;
+};
+
+[[group(0), binding(0)]]
+var org: texture_2d<f32>;
+
+[[group(0), binding(1)]]
+var sampler: sampler;
+
+[[group(0), binding(2)]]
+var<uniform> uniforms: Uniforms;
+
+[[group(0), binding(3)]]
+var up: texture_2d<f32>;
+
+fn quadratic_threshold(color: vec4<f32>, threshold: f32, curve: vec3<f32>) -> vec4<f32> {
+	let br = max(max(color.r, color.g), color.b);
+
+	var rq: f32 = clamp(br - curve.x, 0.0, curve.y);
+	rq = curve.z * rq * rq;
+
+	return color * max(rq, br - threshold) / max(br, 0.0001); 
+}
+
+[[stage(fragment)]]
+fn down_sample_pre_filter(in: VertexOutput) -> [[location(0)]] vec4<f32> {
+	let texel_size = 1.0 / vec2<f32>(textureDimensions(org));
+
+	let scale = texel_size;
+
+	let a = textureSample(org, sampler, in.uv + vec2<f32>(-1.0, -1.0) * scale);
```

Comment:
There’s a lot of texture fetches, which if I understand correctly is one of the more expensive operations. Did you consider any other approaches and if so can you say something about the trade-offs and why you settled on 5is approach?

Reply:
I just went with the implementation detailed in the video. The reasons given there are the following:
"the overlapping further breaks the grid nature of pyramid approaches"
"Works better than comparable sized filters"
"Eliminates pulsating artifacts and temporal stability issues"

That being said, it is probably worth considering implementing a way to choose a cheaper filter if desired.
I personally don't think it's necessary since the filters are being run on comparatively small textures.